[MRG] Make gathering system info easier

blairconrad opened the pull request:
<!--
Please prefix your PR title with [WIP] for PRs that are in progress and change
to [MRG] when you consider them ready for final review.
-->

#### Describe the changes
Emboldened by the comment at the end of https://github.com/pydicom/pydicom/pull/1091#discussion_r419688012, I took it upon myself to suggest an alternative snippet to gather system info for bug reports.

#### Tasks
- [x] Unit tests added that reproduce the issue or prove feature is working
- [x] Fix or feature added
- [ ] Documentation updated (if relevant)
  - [ ] No warnings during build
  - [ ] Preview link (CircleCI -> Artifacts -> `[...]/_build/html/index.html`)
- [ ] Unit tests passing and overall coverage the same or better

blairconrad commented on .github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -19,9 +19,10 @@ How to reproduce the issue. Please include a minimum working code sample, the
+import platform, sys, pydicom
+print(platform.platform(),
+      "\nPython", sys.version,
+      "\npydicom", pydicom.__version__)
```
I switched to 1 print statement from 3 to avoid having the `>>>` s in the output. That's more for the maintainers than the submitter, but I thought it made the results easier to read.

I'd originally put the print on 1 line, but it got up over 80 chars, so I broke it for better rendering.

blairconrad replied:
I toyed a little bit with something like 
```python
import platform
import sys
import importlib

output = [
    platform.platform(),
    "Python " + sys.version,
]

for module in ("pydicom", "numpy", "PIL", "imagecodecs", "gdcm"):
    try:
        m = importlib.import_module(module)
        output.append(module + " " + m.__version__)
    except:
        pass

print("\n".join(output))
```

(in https://gist.github.com/blairconrad/37888634031016865dec93c33ff7b48f), but it's a little heavy, unless you run it from the gist (or other file in the pydicom repo), but that's got its own problems…)

darcymason replied:
> it's a little heavy, unless you run it from the gist (or other file in the pydicom repo), but that's got its own problems…)

Why not make it a file in the pydicom distribution?  I think that could be really useful, might even prevent an issue from being created if people can see that a library is not the version they expected.  

We would still need something in the issue template for a while for those who hadn't updated to pydicom >= 2, but eventually that could become *the* way to gather the info.  "Please run `python -m pydicom.env_info`" (we could work on the name). 

I think every library should be listed though, not just the found ones.  So absence should be indicated:  `gdcm: not installed`, etc.

> If I've crossed the line from "helpful contributor" to "annoying pest", just say. I'm a package maintainer too. I'll back off and not be offended.

Don't worry, you are not there yet.  :wink:

blairconrad replied:
Thanks, @darcymason.

> I think every library should be listed though, not just the found ones. So absence should be indicated: gdcm: not installed, etc

Haha. For fun, I'd actually continued working on the script, and had this sitting around:

```python
import platform
import sys
import importlib

tidbits = [("platform", platform.platform()), ("Python", sys.version)]

for module in ("pydicom", "numpy", "PIL", "jpeg_ls", "gdcm"):
    try:
        m = importlib.import_module(module)
        tidbits.append((module, m.__version__))
    except:
        tidbits.append((module, "not found"))

print(
    "| module | version |\n| --- | --- |\n| "
    + (" |\n| ".join((module + " | " + version for (module, version) in tidbits)) + " | ")
)
```

Output is 
```markdown
| module | version |
| --- | --- |
| platform | Windows-10-10.0.18362-SP0 |
| Python | 3.7.2 (tags/v3.7.2:9a3ffc0492, Dec 23 2018, 23:09:28) [MSC v.1916 64 bit (AMD64)] |
| pydicom | 2.0.0.dev0 |
| numpy | 1.18.3 |
| PIL | 7.1.2 |
| jpeg_ls | not found |
| gdcm | not found |
```

which renders as
| module | version |
| --- | --- |
| platform | Windows-10-10.0.18362-SP0 |
| Python | 3.7.2 (tags/v3.7.2:9a3ffc0492, Dec 23 2018, 23:09:28) [MSC v.1916 64 bit (AMD64)] |
| pydicom | 2.0.0.dev0 |
| numpy | 1.18.3 |
| PIL | 7.1.2 |
| jpeg_ls | not found |
| gdcm | not found |

I can look into adding it as part of the distro, if you're so inclined. Of course, that means there's a lag while people upgrade to the distribution that has the tool (including those Python 2.xers, like you said.)

darcymason replied:
> I can look into adding it as part of the distro, if you're so inclined.

Yes, please.  I really like this - a nice summary table like that is a good debug tool for users, and easy for us to absorb the critical info in an issue.

One coding suggestion, though, to future-proof in case an import module does not have `__version__`, maybe separate the try: except: to just work on the import itself, use an `else:` clause in the try block for the `__version__`.

blairconrad commented on CONTRIBUTING.md:
````diff
@@ -164,15 +164,15 @@ following rules before submitting:
+   print(platform.platform(),
+         "\nPython", sys.version,
+         "\npydicom", pydicom.__version__)
+   ```
````
Conflicts with #1091.

pep8speaks commented:
Hello @blairconrad! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pydicom/env_info.py`](https://github.com/pydicom/pydicom/blob/c2e39616d845704b836bd38d6d7304e5ba3d6c2d/pydicom/env_info.py):

> [Line 6:80](https://github.com/pydicom/pydicom/blob/c2e39616d845704b836bd38d6d7304e5ba3d6c2d/pydicom/env_info.py#L6): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 7:80](https://github.com/pydicom/pydicom/blob/c2e39616d845704b836bd38d6d7304e5ba3d6c2d/pydicom/env_info.py#L7): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)

darcymason commented:
Had a look through - looks good.  Tried it on my system, worked fine.

Any thoughts about unit tests?  I don't think it needs to be exhaustive, perhaps even one test that just confirms the left column values, just to make sure the file is still running in future versions of python.  It would be tricky (and not worth it IMO) to try to test for numpy/no numpy / gdcm or not etc. multiple environments in the pydicom CI test suite, with changing versions over time too.  And the test could call it through `main()`; I don't think it needs to be run as a script for the test.

blairconrad commented:
> Had a look through - looks good. Tried it on my system, worked fine.

Oh, wow. Thanks. I had been keeping in [WIP] to not be a bother.

> Any thoughts about unit tests?

Yeah. I saw the coverage annoyance. I mulling what to do. I agree with everything you said above, regarding simple being enough. Anything fancy is probably just a waste of effort. I'll add some just like you suggest.

While we're talking about this, one thing I've not (yet?) tried - verifying that the versions are reported for gdcm and jpeg_ls. I see `_module not found_`, because I don't have the modules. I don't use them, and have had trouble installing on Windows (even WSL). I was going to try in a GitHub(s) Action later. But if you've independently verified, I could stop.

blairconrad commented:
Added rudimentary unit tests and changed  `**no __version__ attribute**` to `**cannot determine version**` against the possibility that some modules will eventually be referenced that indicate their version by something other than `__version__`.

darcymason commented:
In a python 2 gdcm install on my system, there is no `__version__`.  The best seems to be `GDCM_VERSION`, i.e. ```import gdcm; print(gdcm.GDCM_VERSION)```  I wasn't able yet to get a python 3 version installed.

CharLS/jpeg_ls has no version number of any kind that I can find, except in the setup.py file, which is not available after a normal install.

darcymason commented:
> I wasn't able yet to get a python 3 version installed.

I did manage to get gdcm in python3.  Same result - GDCM_VERSION seems to be the best option.

blairconrad commented:
>  GDCM_VERSION seems to be the best option

Thanks, @darcymason. Amended. Sorry to have put you to that work.

darcymason commented on pydicom/env_info.py:
```diff
@@ -0,0 +1,50 @@
+
+def extract_version(module):
+    if module.__name__ == "gcdm":
+        return getattr(module, "GDCM_VERSION", None)
```
Typo:  gdcm spelled incorrectly.

blairconrad replied:
![the shame](https://media.giphy.com/media/K1QnLV1caRpuw/giphy.gif)

The worst possible typo! Amended. Thanks/sorry.

darcymason replied:
LOL, no problem.  Stuff happens.

blairconrad commented:
Thanks, @darcymason. I had a good time. Sorry about all the back-and-forth I sucked you into. (Don't feel you need to respond to this, as it's just more back-and-forth!)
